src/query/service/src/sessions/session_mgr_metrics.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use databend_common_base::base::ProgressValues;
+use databend_common_metrics::interpreter::METRIC_QUERY_SCAN_PROGRESS_BYTES;
+use databend_common_metrics::interpreter::METRIC_QUERY_SCAN_PROGRESS_ROWS;
+use databend_common_metrics::interpreter::METRIC_QUERY_SPILL_PROGRESS_BYTES;
+use databend_common_metrics::interpreter::METRIC_QUERY_SPILL_PROGRESS_ROWS;
+use databend_common_metrics::interpreter::METRIC_QUERY_WRITE_PROGRESS_BYTES;
+use databend_common_metrics::interpreter::METRIC_QUERY_WRITE_PROGRESS_ROWS;
+use parking_lot::Mutex;
+use prometheus_client::collector::Collector;
+use prometheus_client::encoding::EncodeMetric;
+use prometheus_client::metrics::counter::ConstCounter;
+
+use crate::sessions::SessionManager;
+
+/// [`SessionManagerMetricsCollector`] dumps the progress metrics of scan/write/spills
+/// from the [`SessionManager`]'s running queries to the prometheus. To avoid the progress
+/// metrics being decreased, we also need to accumulate these progress values after the query
+/// is finished.
+#[derive(Clone)]
+pub struct SessionManagerMetricsCollector {
+    inner: Arc<Mutex<SessionManagerMetricsCollectorInner>>,
+}
+
+pub(crate) struct SessionManagerMetricsCollectorInner {
+    session_mgr: Option<Arc<SessionManager>>,
+    finished_scan_total: ProgressValues,
+    finished_write_total: ProgressValues,
+    finished_spill_total: ProgressValues,
+}
+
+impl SessionManagerMetricsCollector {
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(SessionManagerMetricsCollectorInner {
+                session_mgr: None,
+                finished_scan_total: ProgressValues::default(),
+                finished_write_total: ProgressValues::default(),
+                finished_spill_total: ProgressValues::default(),
+            })),
+        }
+    }
+
+    pub fn attach_session_manager(&self, session_mgr: Arc<SessionManager>) {
+        let mut guard = self.inner.lock();
+        guard.session_mgr.replace(session_mgr);
+    }
+
+    pub fn track_finished_query(
+        &self,
+        scan: ProgressValues,
+        write: ProgressValues,
+        join_spill: ProgressValues,
+        aggregate_spill: ProgressValues,
+        group_by_spill: ProgressValues,
+        window_partition_spill: ProgressValues,
+    ) {
+        let mut guard = self.inner.lock();
```

Comment:
Atomic is better?

Reply:
let me have a try

Reply:
tried to use an AtomicPtr instead but finally returned to the Mutex approach 🥲, the exlusive region is not big imo, it only held the lock when query is finished & metrics is scraped..